src/build_vehicle_gui.cpp

```diff
@@ -1409,15 +1444,23 @@ struct BuildVehicleWindow : Window {
 			case VEH_AIRCRAFT:
 				this->GenerateBuildAircraftList();
 				break;
+			}
+
+			this->eng_list.shrink_to_fit();
+			this->eng_list.RebuildDone();
+			this->vscroll->SetCount((uint)this->eng_list.size());
+		}
+
+		if (this->vehicle_type == VEH_TRAIN) {
+			return; // trains should not reach the last sorting
 		}
 
 		this->FilterEngineList();
 
 		_engine_sort_direction = this->descending_sort_order;
 		EngList_Sort(&this->eng_list, _engine_sort_functions[this->vehicle_type][this->sort_criteria]);
 
-		this->eng_list.shrink_to_fit();
-		this->eng_list.RebuildDone();
```

Comment:
The `FilterEngineList()` above some lines above can change the content of the `eng_list`, so I think these lines are needed here.

Reply:
Made some changes that simplify it a lot and keeps most of the existing code structure, thank you!